presets/tuning/text-generation/fine_tuning_api.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import os
+import sys
+from dataclasses import asdict
+from datetime import datetime
+from parser import parse_configs
+
+import torch
+import transformers
+from accelerate import Accelerator
+from dataset import DatasetManager
+from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training
+from transformers import (AutoModelForCausalLM, AutoTokenizer,
+                          BitsAndBytesConfig, HfArgumentParser, Trainer,
+                          TrainingArguments)
+from trl import SFTTrainer
+
+CONFIG_YAML = os.environ.get('YAML_FILE_PATH', '/mnt/config/training_config.yaml')
+
+parsed_configs = parse_configs(CONFIG_YAML)
+
+model_config = parsed_configs.get('ModelConfig')
+tk_params = parsed_configs.get('TokenizerParams')
+bnb_config = parsed_configs.get('QuantizationConfig')
+ext_lora_config = parsed_configs.get('LoraConfig')
+ta_args = parsed_configs.get('TrainingArguments')
+ds_config = parsed_configs.get('DatasetConfig')
+dc_args = parsed_configs.get('DataCollator')
+
+accelerator = Accelerator()
+
+# Load Model Args
+model_args = asdict(model_config)
+if accelerator.distributed_type != "NO":  # Meaning we require distributed training
+    print("Setting device map for distributed training")
+    model_args["device_map"] = {"": accelerator.process_index}
```

Comment:
why device_map has a map with empty key?

Reply:
Empty Key ("") - When the key is an empty string, it means that the entire model is assigned to the specified device. This is a shorthand notation used to simplify the setup.